---
title: "Molecularly Subtyping Embryonal Tumors - Which samples to include?"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Stephanie J. Spielman and Jaclyn Taroni for ALSF CCDL
date: 2019
---

This notebook identifies samples to include in subset files for the purpose of molecularly subtyping embryonal tumors ([`AlexsLemonade/OpenPBTA-analysis#251`](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/251)).

This closed pull request is also relevant to the task at hand: [`AlexsLemonade/OpenPBTA-analysis#401`](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/401).

To summarize - we will identify biospecimen IDs that should be included for downstream analysis using the following criteria:

1. An RNA-seq biospecimen includes a _TTYH1_ fusion (5' partner).
   We can consume the files from the [`fusion-summary` module](../fusion-summary/) for this step.
   (See also: [this comment on `AlexsLemonade/OpenPBTA-analysis#401`](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/401#issuecomment-573669727).)
   
2. An RNA-seq biospecimen sample includes a _MN1_ fusion (5' partner) [per this comment](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/785#issuecomment-695015488).
  Note that the `MN1--PATZ1` fusion is excluded as it is an entity separate of CNS HGNET-MN1 tumors [per this comment](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/788#discussion_r495302880).

3. Any sample with "Supratentorial or Spinal Cord PNET" in the `pathology_diagnosis` column of the metadata `histologies.tsv` (also [per this comment](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/752#issuecomment-697000066)).

4. Any sample with "Neuroblastoma" in the `pathology_diagnosis` column, where `primary_site` does not contain "Other locations NOS", and `pathology_free_text_diagnosis` does not contain "peripheral" or "metastatic".

5. Any sample with "Other" in the `pathology_diagnosis` column of the metadata, and with "embryonal tumor with multilayer rosettes, ros (who grade iv)", "embryonal tumor, nos, congenital type", "ependymoblastoma" or "medulloepithelioma" in the `pathology_free_text_diagnosis` column [per this comment](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/752#issuecomment-697000066).

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/molecular-subtyping-embryonal/01-samples-to-subset.Rmd', clean = TRUE)"
```

## Set Up

### Libraries and functions

```{r}
library(tidyverse)
```

### Directories

```{r}
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")

results_dir <- "results"
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

subset_dir <- "subset-files"
```

### Read in relevant files

```{r message=FALSE}
histologies_df <- read_tsv(file.path(data_dir, "histologies-base.tsv"), 
                           guess_max = 100000)
fusion_summary_df <- read_tsv(file.path(data_dir, 
                                        "fusion_summary_embryonal_foi.tsv"))
arriba <- read_tsv(file.path(data_dir, "fusion-arriba.tsv.gz"))
starfusion <- read_tsv(file.path(data_dir, "fusion-starfusion.tsv.gz"))
dgd_fusion <- read_tsv(file.path(data_dir, "fusion-dgd.tsv.gz"))

# Read in the JSON file that contains the strings we'll use to include or
# exclude samples for subtyping - see 00-embryonal-select-pathology-dx.Rmd
path_dx_list <- jsonlite::fromJSON(
  file.path(subset_dir, 
            "embryonal_subtyping_path_dx_strings.json")
)
```

### Output file

```{r}
output_file <- file.path(results_dir, "biospecimen_ids_embryonal_subtyping.tsv")
methly_output_file <- file.path(results_dir, "methyl_embryonal_subtyping.tsv")
```

## Identify relevant samples

### _TTYH1_ fusions

We'll use the fusion summary file to find samples that contain _TTYH1_.
We're only interested in fusions where _TTYH1_ is the 5' fusion partner.

```{r}
ttyh1_fusions <- fusion_summary_df %>%
  select(Kids_First_Biospecimen_ID, starts_with("TTYH1"))
```

Which samples have these fusions?

```{r}
ttyh1_fusion_biospecimens <- ttyh1_fusions %>%
  # add a column that counts how many TTYH1 fusions are present
  mutate(fusion_count = rowSums(select(., starts_with("TTYH1")))) %>%
  # if there's any TTYH1 fusion - pull out the biospecimen ID
  filter(fusion_count > 0) %>%
  pull(Kids_First_Biospecimen_ID)
```

make sure that at least one TTYH1 fusion breakpoint is in only the C19MC miRNA 
cluster genomic region: `chr19:53,627,063-53,788,978` to keep the sample
```{r}
# TTYHI fusions in ariba
arriba_ttyh1_fusions <- arriba %>% 
  filter(str_detect(`gene1--gene2`, "^TTYH1"), 
         str_detect(breakpoint1, "^chr19"), 
         str_detect(breakpoint2, "^chr19"))
if (length(arriba_ttyh1_fusions$tumor_id) > 0 ) {
  ttyh1_fusions <- arriba_ttyh1_fusions %>% 
    select(breakpoint1, breakpoint2, tumor_id) %>% 
    mutate(breakpoint1 = str_split(breakpoint1, ":")[[1]][2], 
           breakpoint2 = str_split(breakpoint2, ":")[[1]][2]) %>%  
    filter(between(as.integer(breakpoint1), 53627062, 53788979) | 
             between(as.integer(breakpoint2), 53627062, 53788979)) %>% 
    pull(tumor_id) %>% unique()
}

# v12 starfusion does not have TTYHI fusions where _TTYH1_ is the 5' fusion partner
starfusion_ttyh1_fusions <- starfusion %>% 
  filter(str_detect(FusionName, "^TTYH1"), 
         str_detect(LeftBreakpoint, "^chr19"), 
         str_detect(RightBreakpoint, "^chr19"))
if (length(starfusion_ttyh1_fusions$tumor_id) > 0 ) {
  starfusion_ttyh1_fusions <- starfusion_ttyh1_fusions %>% 
    select(LeftBreakpoint, RightBreakpoint, tumor_id) %>% 
    mutate(LeftBreakpoint = str_split(LeftBreakpoint, ":")[[1]][2], 
           RightBreakpoint = str_split(RightBreakpoint, ":")[[1]][2]) %>%  
    filter(between(as.integer(LeftBreakpoint), 53627062, 53788979) | 
             between(as.integer(RightBreakpoint), 53627062, 53788979)) %>% 
    pull(tumor_id) %>% unique()
  ttyh1_fusions <- c(starfusion_ttyh1_fusions, ttyh1_fusions)
}

# v12 DGD fusion does not have TTYHI fusions where _TTYH1_ is the 5' fusion partner
dgd_ttyh1_fusions <- dgd_fusion %>% 
  filter(str_detect(FusionName, "^TTYH1"),
         str_detect(LeftBreakpoint, "^chr19"),
         str_detect(RightBreakpoint, "^chr19"))
if (length(dgd_ttyh1_fusions$tumor_id) > 0 ) {
  dgd_ttyh1_fusions <- dgd_ttyh1_fusions %>% 
    select(LeftBreakpoint, RightBreakpoint, tumor_id) %>% 
    mutate(LeftBreakpoint = str_split(LeftBreakpoint, ":")[[1]][2], 
           RightBreakpoint = str_split(RightBreakpoint, ":")[[1]][2]) %>%  
    filter(between(as.integer(LeftBreakpoint), 53627062, 53788979) | 
             between(as.integer(RightBreakpoint), 53627062, 53788979)) %>% 
    pull(tumor_id) %>% unique()
  ttyh1_fusions <- c(dgd_ttyh1_fusions, ttyh1_fusions)
}

# keep samples with with at least one break point within 
# genomic region: `chr19:53,627,063-53,788,978` 
ttyh1_fusion_biospecimens <-  ttyh1_fusion_biospecimens %in% ttyh1_fusions
```

### _MN1_ fusions

Now, we'll use the fusion summary file to find samples that contain _MN1_.
We're only interested in fusions where _MN1_ is the 5' fusion partner.

```{r}
# We will also filter out the specific _MN1--PATZ1_ fusion based on github comment: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/788#discussion_r495212879
mn1_fusion_biospecimens <- fusion_summary_df %>%
  select(Kids_First_Biospecimen_ID, starts_with("MN1"), -("MN1--PATZ1"))
```

Which samples have these fusions?

```{r}
fusion_biospecimens <- mn1_fusion_biospecimens %>%
  # add a column that counts how many MN1 fusions are present
  mutate(fusion_count = rowSums(select(., starts_with("MN1")))) %>%
  # if there's any MN1 fusion - pull out the biospecimen ID
  filter(fusion_count > 0) %>%
  pull(Kids_First_Biospecimen_ID) %>%
  # Add the TTYH1 fusion biospecimens to this object
  c(ttyh1_fusion_biospecimens)
```

DNA-seq biospecimens associated with the samples that contain the relevant fusions.

```{r}
relevant_samples <- histologies_df %>%
  filter(Kids_First_Biospecimen_ID %in% fusion_biospecimens) %>%
  pull(sample_id)

relevant_sample_df <- histologies_df %>%
  filter(sample_id %in% relevant_samples,
         experimental_strategy %in% c("WGS", "WXS", "Targeted Sequencing"),
         sample_type == "Tumor",
         is.na(RNA_library))

relevant_sample_df
```

We've captured a medulloblastoma tumor and a pineoblastoma tumor with the _TTYH1_ fusion criterion, and a low-grade glioma tumor with the _MN1_ fusion criterion.

```{r}
fusion_biospecimens <- c(
  fusion_biospecimens,
  relevant_sample_df %>% pull(Kids_First_Biospecimen_ID)
)
```

### Disease labels

First subset to solid **tumors**, excluding any derived cell lines.

```{r}
# removed  filter composition == "Solid Tissue"
embryonal_df <- histologies_df %>%
  filter(sample_type == "Tumor")
```

Now filter to include and exclude using the `pathology_diagnosis` and `pathology_free_text_diagnosis` labels.

First, we will create an object filtered to include only samples with exact matches for the `path_dx_terms` (which is currently "Supratentorial or Spinal Cord PNET") in the `pathology_diagnosis` column.

```{r}
pnet_df <- embryonal_df %>%
  # Filter for exact match of the defined pathology diagnosis term in `path_dx_list$include_path_dx`
  filter(pathology_diagnosis %in% path_dx_list$include_path_dx)
```

Now we will create an objected filtered to include any sample with "Other" in the `pathology_diagnosis` column of the metadata, and with the `free_text_dx_terms` in the `pathology_free_text_diagnosis` column.

```{r}
pathology_free_text_df <- embryonal_df %>%
  # Filter for exact matches of `pathology_diagnosis == "Other"` and defined pathology free text diagnosis terms in `free_text_dx_terms`
  filter(pathology_diagnosis == "Other",
         pathology_free_text_diagnosis %in% path_dx_list$include_free_text) 
```

Any sample with "Neuroblastoma" in the `pathology_diagnosis` column, where `primary_site` does not contain "Other locations NOS", and `pathology_free_text_diagnosis` does not contain "peripheral" or "metastatic".

```{r}
neuroblastoma_df <- embryonal_df %>%
  filter(
    pathology_diagnosis == "Neuroblastoma",
    str_detect(str_to_lower(primary_site), "other locations nos", negate = TRUE),
    str_detect(
      str_to_lower(pathology_free_text_diagnosis),
      "peripheral|metastatic",
      negate = TRUE
    )
  )
```

Now let's bind all of the information in our filtered data frames together into one object.

```{r}
# Binding all of the rows
filtered_embryonal_df <- bind_rows(pnet_df,
                                   pathology_free_text_df,
                                   neuroblastoma_df)
filtered_embryonal_df
```

### Combine biospecimen IDs

Now let's combine biospecimen IDs using the fusion information and the filtered clinical file.

```{r}
all_biospecimen_ids <- 
  unique(c(filtered_embryonal_df %>% pull(Kids_First_Biospecimen_ID),
           fusion_biospecimens))
```

Write to file.

```{r}
histologies_df %>%
  dplyr::filter(Kids_First_Biospecimen_ID %in% all_biospecimen_ids) %>%
  dplyr::filter(!cohort %in% c("GMKF", "TARGET")) %>% 
  dplyr::select(Kids_First_Biospecimen_ID) %>%
  write_tsv(output_file)
```

### Methylation embryonal subtypes

Write to file high confidence methylation classification subtypes

```{r}
# high confidence methylation classification subtypes
emb_methyl_subtyped <- histologies_df %>% 
  dplyr::filter(pathology_diagnosis %in% path_dx_list$include_path_dx |
                  pathology_free_text_diagnosis %in% path_dx_list$include_free_text,
                dkfz_v12_methylation_subclass_score >= 0.8,
                experimental_strategy == "Methylation", 
                sample_type == "Tumor", 
                composition == "Solid Tissue",
                cohort != "TARGET") %>% 
  dplyr::mutate(molecular_subtype_methyl = 
                  case_when(dkfz_v12_methylation_subclass == "CNS_NB_FOXR2" ~ "CNS NB-FOXR2",
                            dkfz_v12_methylation_subclass == "ETMR_C19MC" ~ "ETMR, C19MC-altered",
                            dkfz_v12_methylation_subclass == "CNS_BCOR_ITD" ~ "CNS BCOR",
                            dkfz_v12_methylation_subclass == "NET_CXXC5" ~ "CNS HGNET-MN1",
                            dkfz_v12_methylation_subclass == "ABM_MN1" ~ "CNS HGNET-MN1")) %>% 
  dplyr::select(Kids_First_Biospecimen_ID, Kids_First_Participant_ID, sample_id, molecular_subtype_methyl) %>%
  dplyr::rename("Kids_First_Biospecimen_ID_Methyl" = "Kids_First_Biospecimen_ID") %>% 
  dplyr::filter(!is.na(molecular_subtype_methyl)) %>% 
  dplyr::distinct()
  
# low confidence methylation classification subtypes
emb_methyl_not_subtyped <- histologies_df %>% 
  dplyr::filter(pathology_diagnosis %in% path_dx_list$include_path_dx |
                  pathology_free_text_diagnosis %in% path_dx_list$include_free_text,
                !Kids_First_Biospecimen_ID %in% unique(emb_methyl_subtyped$Kids_First_Biospecimen_ID_Methyl),
                experimental_strategy == "Methylation", 
                sample_type == "Tumor", 
                composition == "Solid Tissue",
                cohort != "TARGET") %>%
  dplyr::left_join(emb_methyl_subtyped %>% dplyr::select(Kids_First_Participant_ID, sample_id, molecular_subtype_methyl),
                   by = c("Kids_First_Participant_ID", "sample_id")) %>% 
  dplyr::select(Kids_First_Biospecimen_ID, Kids_First_Participant_ID, sample_id, molecular_subtype_methyl) %>%
  dplyr::rename("Kids_First_Biospecimen_ID_Methyl" = "Kids_First_Biospecimen_ID") %>% 
  dplyr::distinct()

# merge methyl and write to file
dplyr::bind_rows(emb_methyl_subtyped, emb_methyl_not_subtyped) %>% 
  readr::write_tsv(methly_output_file)
```

## Session Info

```{r}
sessionInfo()
```
